src/services/codefixes/importFixes.ts

```diff
@@ -504,45 +504,114 @@ namespace ts.codefix {
+                                        if (mainExportFile === getCanonicalFileName(path)) {
+                                            return packageRootPath;
                                         }
                                     }
                                 }
                             }
-                            catch (e) { }
+
+                            // We still have a file name - remove the extension
+                            const fullModulePathWithoutExtension = removeFileExtension(path);
+
+                            // If the file is /index, it can be imported by its directory name
+                            if (getCanonicalFileName(fullModulePathWithoutExtension.substring(parts.fileNameIndex)) === "/index") {
+                                return fullModulePathWithoutExtension.substring(0, parts.fileNameIndex);
+                            }
+
+                            return fullModulePathWithoutExtension;
                         }
 
-                        return getPackageNameFromAtTypesDirectory(relativeFileName);
+                        function getNodeResolvablePath(path: string): string {
+                            const basePath = path.substring(0, parts.nodeModulesIndex);
+                            if (sourceDirectory.indexOf(basePath) === 0) {
+                                // if node_modules folder is in this folder or any of its parent folders, no need to keep it.
+                                return path.substring(parts.topLevelPackageNameIndex + 1);
+                            }
+                            else {
+                                return getRelativePath(path, sourceDirectory);
+                            }
+                        }
                     }
                 }
 
+                function getNodeModulePathParts(fullPath: string) {
+                    // If fullPath can't be valid module file within node_modules, returns undefined.
+                    // Example of expected pattern: /base/path/node_modules/[otherpackage/node_modules/]package/[subdirectory/]file.js
+                    // Returns indices:                       ^            ^                                   ^             ^
+
+                    let nodeModulesIndex = 0;
+                    let topLevelPackageNameIndex = 0;
+                    let packageRootIndex = 0;
+                    let fileNameIndex = 0;
+
+                    enum States {
+                        BeforeNodeModules,
+                        NodeModules,
+                        PackageContent
+                    }
+
+                    let partStart = 0;
+                    let partEnd = 0;
+                    let state = States.BeforeNodeModules;
+
+                    while (partEnd >= 0) {
+                        partStart = partEnd;
+                        partEnd = fullPath.indexOf("/", partStart + 1);
+                        const part = fullPath.substring(partStart, partEnd);

```

Comment:
If the goal of doing everything with indices with indices was to avoid creating temporary strings, this should probably be created on demand (e.g. never in `States.NodeModules`) or, better still, "/node_modules" could be detected in-place.